File: 期末/上机/02_MATLAB算法模板/01_单步法/T01_Taylor_Series.m
%% Taylor 级数法模板：y_{n+1}=y_n+sum_{j=1}^p h^j/j! * y^(j)(t_n)
clear; clc; close all;                                      % MATLAB 语法：清变量、清命令窗、关图窗，避免旧数据干扰。
fDeriv = {@(t,y) -y+t.^2+1, @(t,y) y-t.^2+2*t-1};         % 【必须替换】依次填写 y'、y''、...；花括号创建 cell 数组。
t0 = 0; T = 1; y0 = 1; h = 0.1;                           % 【必须替换】左端点、右端点、初值、固定步长。
p = numel(fDeriv);                                         % numel 返回导数函数个数，也就是 Taylor 方法阶数 p。
N = round((T-t0)/h);                                       % round 将步数取整；本模板要求 (T-t0)/h 为整数。
assert(abs(t0+N*h-T)<1e-12,'步长必须整除区间长度。');       % assert 在条件不成立时停止，防止终点与题目不一致。
t = t0+(0:N)*h;                                            % 冒号语法 0:N 生成整数网格，向量化得到所有节点。
y = zeros(1,N+1);                                          % zeros 预分配内存；N 步包含 N+1 个节点。
y(1) = y0;                                                 % MATLAB 下标从 1 开始，所以 y(1) 对应数学上的 y_0。
for n = 1:N                                                % for 循环逐步推进；每步仅使用当前点信息，属于单步法。
    increment = 0;                                         % 累积本步 Taylor 增量，先置零。
    for j = 1:p                                            % 依次加入 h*y'、h^2/2!*y'' 等项。
        increment = increment + h^j/factorial(j)*fDeriv{j}(t(n),y(n)); % cell 用 {j} 取函数；factorial(j)=j!。
    end                                                    % 结束内层 Taylor 求和。
    y(n+1) = y(n)+increment;                               % 截断到 p 阶导数，局部截断误差通常为 O(h^(p+1))。
end                                                        % 结束时间推进。
disp(table(t.',y.','VariableNames',{'t','Taylor_y'}));      % 点转置 .' 把行向量变列向量；table 生成结果表。
plot(t,y,'o-','LineWidth',1.3); grid on;                    % plot 画数值解；grid on 打开网格便于观察。
xlabel('t'); ylabel('y'); title(sprintf('%d 阶 Taylor 方法',p)); % sprintf 把阶数写入标题。

File: 期末/上机/03_Python算法模板/01_单步法/T01_Taylor_Series.py
# Taylor 级数法：y_{n+1}=y_n+sum_{j=1}^p h^j/j!*y^(j)(t_n)。  # 数值分析：截断到 p 阶导数时通常得到 p 阶方法。
import math  # Python 语法：导入标准库 math，用 factorial 计算阶乘。
import numpy as np  # Python 语法：给 NumPy 起别名 np，用于数组与等距网格。
import matplotlib.pyplot as plt  # Python 语法：导入绘图库 pyplot，并使用别名 plt。

derivatives = [lambda t, y: -y + t**2 + 1, lambda t, y: y - t**2 + 2*t - 1]  # 【必须替换】依次填写 y'、y''、...；lambda 创建匿名函数。
t0, T, y0, h = 0.0, 1.0, 1.0, 0.1  # 【必须替换】左端点、右端点、初值、固定步长；逗号可一次赋多个变量。
p = len(derivatives)  # Python 语法：len 返回列表长度；数值分析：导数函数个数就是 Taylor 方法阶数 p。
N = round((T-t0)/h)  # Python 语法：round 取最近整数；固定步长方法需要整数步数。
assert abs(t0+N*h-T) < 1e-12, "步长必须整除区间长度。"  # Python 语法：assert 条件失败时停止，防止实际终点错误。
t = t0 + np.arange(N+1)*h  # NumPy 语法：arange(N+1) 生成 0,...,N，再向量化得到全部节点。
y = np.zeros(N+1, dtype=float)  # NumPy 语法：zeros 预分配浮点数组；N 步需要 N+1 个节点值。
y[0] = y0  # Python 下标从 0 开始，因此 y[0] 对应数学记号 y_0。
for n in range(N):  # Python 语法：range(N) 生成 0,...,N-1；数值分析：逐步从 t_n 推进到 t_{n+1}。
    increment = 0.0  # 本步 Taylor 增量先置零。
    for j, derivative in enumerate(derivatives, start=1):  # enumerate 同时给出阶数 j 和第 j 阶导数函数。
        increment += h**j/math.factorial(j)*derivative(t[n], y[n])  # 数值分析：累加 h^j/j!*y^(j)(t_n)。
    y[n+1] = y[n] + increment  # Taylor 截断公式完成一步推进，局部截断误差通常为 O(h^(p+1))。
for tn, yn in zip(t, y):  # Python 语法：zip 把时间与数值解逐项配对，方便输出表格。
    print(f"t={tn:.3f}, Taylor_y={yn:.10f}")  # Python 语法：f-string 控制小数位并插入变量。
plt.plot(t, y, "o-", linewidth=1.3, label=f"Taylor order {p}")  # 绘图语法：o- 表示节点圆点加连线。
plt.xlabel("t")  # 绘图语法：设置横轴名称。
plt.ylabel("y")  # 绘图语法：设置纵轴名称。
plt.grid(True)  # 绘图语法：打开网格，便于观察数值变化。
plt.legend()  # 绘图语法：显示图例。
plt.tight_layout()  # 绘图语法：自动调整边距，防止文字被裁切。
plt.show()  # 绘图语法：显示图像；批量验证时使用 Agg 后端，不会弹窗。
